cmd/sgx_plugin/sgx_plugin.go

```diff
@@ -0,0 +1,133 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"path"
+	"runtime"
+	"strconv"
+
+	dpapi "github.com/intel/intel-device-plugins-for-kubernetes/pkg/deviceplugin"
+	"k8s.io/klog"
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+const (
+	// Device plugin settings.
+	namespace              = "sgx.intel.com"
+	deviceTypeEnclave      = "enclave"
+	deviceTypeProvision    = "provision"
+	devicePath             = "/dev"
+	podsPerCoreEnvVariable = "PODS_PER_CORE"
+	defaultPodsPerCore     = 10
+)
+
+type devicePlugin struct {
+	devfsDir   string
+	nEnclave   uint
+	nProvision uint
+	scanDone   chan bool
+}
+
+func newDevicePlugin(devfsDir string, nEnclave, nProvision uint) *devicePlugin {
+	return &devicePlugin{
+		devfsDir:   devfsDir,
+		nEnclave:   nEnclave,
+		nProvision: nProvision,
+		scanDone:   make(chan bool, 1),
+	}
+}
+
+func (dp *devicePlugin) Scan(notifier dpapi.Notifier) error {
+	devTree, err := dp.scan()
+	if err != nil {
+		return err
+	}
+	notifier.Notify(devTree)
+
+	// Wait forever to prevent manager run loop from exiting.
+	<-dp.scanDone
+	return nil
+}
+
+func (dp *devicePlugin) scan() (dpapi.DeviceTree, error) {
+	devTree := dpapi.NewDeviceTree()
+
+	// Assume that both /dev/sgx/enclave and /dev/sgx/provision must be present.
+	sgxEnclavePath := path.Join(dp.devfsDir, "sgx", "enclave")
+	sgxProvisionPath := path.Join(dp.devfsDir, "sgx", "provision")
+	if _, err := os.Stat(sgxEnclavePath); err != nil {
+		klog.Error("No SGX enclave file available: ", err)
+		return devTree, nil
+	}
+	if _, err := os.Stat(sgxProvisionPath); err != nil {
+		klog.Error("No SGX provision file available: ", err)
+		return devTree, nil
+	}
+
+	for i := uint(0); i < dp.nEnclave; i++ {
+		devID := fmt.Sprintf("%s-%d", "sgx-enclave", i)
+		nodes := []pluginapi.DeviceSpec{{HostPath: sgxEnclavePath, ContainerPath: sgxEnclavePath, Permissions: "rw"}}
+		devTree.AddDevice(deviceTypeEnclave, devID, dpapi.NewDeviceInfo(pluginapi.Healthy, nodes, nil, nil))
+	}
+	for i := uint(0); i < dp.nProvision; i++ {
+		devID := fmt.Sprintf("%s-%d", "sgx-provision", i)
+		nodes := []pluginapi.DeviceSpec{{HostPath: sgxProvisionPath, ContainerPath: sgxProvisionPath, Permissions: "rw"}}
+		devTree.AddDevice(deviceTypeProvision, devID, dpapi.NewDeviceInfo(pluginapi.Healthy, nodes, nil, nil))
+	}
+	return devTree, nil
+}
+
+func getDefaultPodCount(nCPUs uint) uint {
+	// By default we provide as many enclave resources as there can be pods
+	// running on the node. The problem is that this value is configurable
+	// either via "--pods-per-core" or "--max-pods" kubelet options. We get the
+	// limit by multiplying the number of cores in the system with env variable
+	// "PODS_PER_CORE".
+
+	podsPerCore := uint(defaultPodsPerCore)
```

Comment:
Probably it would be safer to use a typed global constant for `defaultPodsPerCore`.

Reply:
Ok, fixed.